File: fix-svg-icons.mjs
/**
 * fix-svg-icons.mjs  —  Run from your project root:
 *   node fix-svg-icons.mjs
 *
 * Converts every SVG-disguised-as-.png in src/assets/tech/ and
 * src/assets/icons/ into a real 256x256 PNG so Three.js can load them.
 *
 * Uses `sharp` which works on Windows without any native DLLs.
 */

import { execSync } from 'child_process';
import { readFileSync, writeFileSync, readdirSync, existsSync, statSync } from 'fs';
import { join, resolve } from 'path';
import { fileURLToPath } from 'url';

// ── Make sure sharp is available ──────────────────────────────────────────
let sharp;
try {
  sharp = (await import('sharp')).default;
  console.log('✅ sharp already installed\n');
} catch {
  console.log('📦 Installing sharp (one-time setup)...');
  execSync('npm install --save-dev sharp', { stdio: 'inherit' });
  sharp = (await import('sharp')).default;
  console.log('');
}

// ── Folders to scan ───────────────────────────────────────────────────────
const DIRS = ['src/assets/tech', 'src/assets/icons'];
const SIZE = 256;

let converted = 0;
let skipped   = 0;
let failed    = 0;

for (const folder of DIRS) {
  const abs = resolve(folder);

  if (!existsSync(abs)) {
    console.log(`⚠️  Skipping (not found): ${folder}`);
    continue;
  }

  console.log(`\n📁 Scanning: ${folder}`);
  console.log('─'.repeat(50));

  const files = readdirSync(abs).filter(f => f.endsWith('.png'));

  if (files.length === 0) {
    console.log('   (no .png files found)');
    continue;
  }

  for (const filename of files) {
    const filepath = join(abs, filename);
    const raw = readFileSync(filepath);

    // ── Detect SVG by content, not extension ──────────────────────────
    const header = raw.slice(0, 1024).toString('utf8');
    const isSVG  =
      header.includes('<svg') ||
      header.includes('<?xml') ||
      header.includes('xmlns="http://www.w3.org/2000/svg"');

    // Also check PNG magic bytes: first 8 bytes = 89 50 4E 47 0D 0A 1A 0A
    const isPNG =
      raw[0] === 0x89 &&
      raw[1] === 0x50 &&
      raw[2] === 0x4e &&
      raw[3] === 0x47;

    if (isPNG) {
      console.log(`   ✓  Already real PNG: ${filename}  (${raw.length} bytes)`);
      skipped++;
      continue;
    }

    if (!isSVG) {
      console.log(`   ⚠️  Unknown format, skipping: ${filename}`);
      skipped++;
      continue;
    }

    const sizeBefore = raw.length;
    process.stdout.write(`   🔄 Converting SVG→PNG: ${filename} ... `);

    try {
      const pngBuffer = await sharp(raw, { density: 192 })
        .resize(SIZE, SIZE, {
          fit: 'contain',
          background: { r: 0, g: 0, b: 0, alpha: 0 }, // transparent bg
        })
        .png({ compressionLevel: 9 })
        .toBuffer();

      writeFileSync(filepath, pngBuffer);
      console.log(`✅  (${sizeBefore} → ${pngBuffer.length} bytes)`);
      converted++;
    } catch (err) {
      console.log(`❌  ERROR: ${err.message}`);
      failed++;
    }
  }
}

// ── Summary ───────────────────────────────────────────────────────────────
console.log('\n' + '='.repeat(50));
console.log(`✅  Converted : ${converted} files`);
console.log(`⏭️   Skipped   : ${skipped} files`);
if (failed > 0) {
  console.log(`❌  Failed    : ${failed} files`);
  console.log('\nFor failed files, manually download a PNG version from:');
  console.log('  → https://devicons.dev');
  console.log('  → https://simpleicons.org');
}
console.log('\n▶  Now restart your dev server:');
console.log('   npm run dev');
console.log('='.repeat(50) + '\n');

File: fix-svg-icons.py
#!/usr/bin/env python3
"""
fix-svg-icons.py
Run from your project root: python fix-svg-icons.py

Finds every .png file in src/assets/tech/ and src/assets/icons/
that is actually an SVG, and re-saves it as a real 256x256 PNG.
Three.js Ball canvas requires real raster images — SVGs won't work.
"""

import os
import sys

try:
    import cairosvg
except ImportError:
    print("Installing cairosvg...")
    os.system(f"{sys.executable} -m pip install cairosvg")
    import cairosvg

DIRS = [
    "src/assets/tech",
    "src/assets/icons",
    "src/assets/projects",
]

SIZE = 256
converted = 0
skipped = 0

for folder in DIRS:
    if not os.path.exists(folder):
        print(f"  ⚠️  Folder not found, skipping: {folder}")
        continue

    for filename in os.listdir(folder):
        if not filename.endswith(".png"):
            continue

        filepath = os.path.join(folder, filename)

        # Read raw bytes to check if it's actually an SVG
        with open(filepath, "rb") as f:
            header = f.read(512)

        is_svg = (
            b"<svg" in header
            or b"<?xml" in header
            or b"xmlns" in header
        )

        if not is_svg:
            print(f"  ✓  Already a real PNG: {filename}")
            skipped += 1
            continue

        print(f"  🔄 Converting SVG→PNG: {filename} ...", end=" ")
        try:
            with open(filepath, "rb") as f:
                svg_data = f.read()

            png_data = cairosvg.svg2png(
                bytestring=svg_data,
                output_width=SIZE,
                output_height=SIZE,
            )

            with open(filepath, "wb") as f:
                f.write(png_data)

            print("✅")
            converted += 1
        except Exception as e:
            print(f"❌  Error: {e}")

print()
print("=" * 50)
print(f"✅  Converted: {converted} files")
print(f"⏭️   Skipped (already PNG): {skipped} files")
print()
print("Restart your Next.js dev server: npm run dev")
print("=" * 50)